lib/spack/external/py2/typing.py

```diff
@@ -1,28 +1,30 @@
-# Copyright 2013-2020 Lawrence Livermore National Security, LLC and other
+# Copyright 2013-2021 Lawrence Livermore National Security, LLC and other
 # Spack Project Developers. See the top-level COPYRIGHT file for details.
 #
 # SPDX-License-Identifier: (Apache-2.0 OR MIT)
 """
 This is a fake set of symbols to allow spack to import typing in python
 versions where we do not support type checking (<3)
 """
+from collections import defaultdict
+
 Annotated = None
 Any = None
-Callable = None
+Callable = defaultdict(lambda: object)
```

Comment:
I like this workaround for being able to declare re-usable composite types.  Would you mind looking at a PR to finish this for the common types here?  It's a lot cleaner than what I'd been contemplating, and should help a lot with some of the nasty union types we have to deal with.

Reply:
Yes absolutely! I think it would go best in #21720. I'm actually quite proud of this because I know the pex project just decided not to use `Union` or anything in its mypy migration because we (including me) hadn't figured it out yet.